lighthouse-core/audits/bootup-time.js

```diff
@@ -0,0 +1,203 @@
+  'JS Frame': group.scripting,
+  'Input Latency': group.scripting,
+  'Layout': group.styleLayout,
+  'Major GC': group.scriptGC,
+  'DOMContentLoaded event': group.scripting,
+  'First paint': group.painting,
+  'FMP': group.painting,
+  'FMP candidate': group.painting,
+  'Load event': group.scripting,
+  'Minor GC': group.scriptGC,
+  'Paint': group.painting,
+  'Paint Image': group.images,
+  'Paint Setup': group.painting,
+  'Parse Stylesheet': group.parseHTML,
+  'Parse HTML': group.parseHTML,
+  'Parse Script': group.scriptParseCompile,
+  'Other': group.other,
+  'Rasterize Paint': group.painting,
+  'Recalculate Style': group.styleLayout,
+  'Request Animation Frame': group.scripting,
+  'Request Idle Callback': group.scripting,
+  'Request Main Thread Frame': group.painting,
+  'Image Resize': group.images,
+  'Finish Loading': group.loading,
+  'Receive Data': group.loading,
+  'Receive Response': group.loading,
+  'Send Request': group.loading,
+  'Run Microtasks': group.scripting,
+  'Schedule Style Recalculation': group.styleLayout,
+  'Scroll': group.compositing,
+  'Task': group.other,
+  'Timer Fired': group.scripting,
+  'Install Timer': group.scripting,
+  'Remove Timer': group.scripting,
+  'Timestamp': group.scripting,
+  'Update Layer': group.compositing,
+  'Update Layer Tree': group.compositing,
+  'User Timing': group.scripting,
+  'Create WebSocket': group.scripting,
+  'Destroy WebSocket': group.scripting,
+  'Receive WebSocket Handshake': group.scripting,
+  'Send WebSocket Handshake': group.scripting,
+  'XHR Load': group.scripting,
+  'XHR Ready State Change': group.scripting,
+};
+
+class BootupTime extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Performance',
+      name: 'bootup-time',
+      description: 'JavaScript boot-up time is high (> 4s)',
+      failureDescription: 'JavaScript boot-up time is too high.',
+      helpText: 'Consider reducing the time spent parsing, compiling and executing JS. ' +
+        'You may find delivering smaller JS payloads helps with this.',
+      requiredArtifacts: ['traces'],
+    };
+  }
+
+  /**
+   * @param {!Array<TraceEvent>=} trace
+   * @return {!Map<string, Number>}
+   */
+  static getExecutionTimingsByURL(trace) {
+    const timelineModel = new DevtoolsTimelineModel(trace);
+    const bottomUpByName = timelineModel.bottomUpGroupBy('URL');
+    const result = new Map();
+
+    bottomUpByName.children.forEach((perUrlNode, url) => {
+      // when url is "" or about:blank, we skip it
+      if (!url || url === 'about:blank') {
+        return;
+      }
+
+      const tasks = {};
+      perUrlNode.children.forEach((perTaskPerUrlNode) => {
+        const taskGroup = WebInspector.TimelineUIUtils.eventStyle(perTaskPerUrlNode.event);
+        tasks[taskGroup.title] = tasks[taskGroup.title] || 0;
+        tasks[taskGroup.title] += Number((perTaskPerUrlNode.selfTime || 0).toFixed(1));
```

Comment:
i felt weird about this rounding here as we're summing, so i moved it to outside of the loop.